kernel/src/process_standard.rs

```diff
@@ -1505,8 +1505,17 @@ impl<C: 'static + Chip> ProcessStandard<'_, C> {
         // Initial size of the kernel-owned part of process memory can be
         // calculated directly based on the initial size of all kernel-owned
         // data structures.
-        let initial_kernel_memory_size =
-            grant_ptrs_offset + Self::CALLBACKS_OFFSET + Self::PROCESS_STRUCT_OFFSET;
+        //
+        // We require our kernel memory break (located at the end of the
+        // MPU-returned allocated memory region) to be word-aligned. However, we
+        // don't have any explicit alignment constraints from the MPU. To ensure
+        // that the below kernel-owned data structures still fit into the
+        // kernel-owned memory even with padding for alignment, add an extra
+        // `sizeof(usize)` bytes.
+        let initial_kernel_memory_size = grant_ptrs_offset
+            + Self::CALLBACKS_OFFSET
+            + Self::PROCESS_STRUCT_OFFSET
+            + core::mem::size_of::<usize>();
```

Comment:
Why not just calculate the exact needed padding?

Reply:
We don't know this ahead of time, it's up to the MPU where the end of the allocation is placed. We have three options:
- Add an additional "allocation end" alignment constraint to `allocate_app_memory_region`, where we can then specify that we want the end of the region word-size aligned. This seems sub-optimal, as this is not even a kernel-invariant, but rather one of `ProcessStandard`.
- Add `core::mem::size_of::<usize>() - 1` to the initial kernel memory size, which is the maximum amount of padding we'll ever need. However, with our current MPU implementations, this will mean that we'll *always* end on an unaligned address, and always waste those 3 bytes.
- Add `core::mem::size_of::<usize>()`. This is one byte more than required, but it would mean that an MPU implementation which naturally aligns regions to a word boundary places this padding also on a word boundary. We would have a slightly larger `initial_kernel_memory_size` than required, but wouldn't waste any of this memory, for instance when we do grant allocations.

Reply:
We have precedent for the first option with `grant`: we request both the size and the alignment when creating a grant. I think it's reasonable to say that alignment is a fundamental operation for allocation.

But I'm not sure I feel strongly about changing the interface for this purpose.

Reply:
I'm all for adding alignment to the interface, but what feels weird is to add an alignment requirement towards the _end_ of a region. Usually the MPU performs alignment inserting padding in the front. In this case, we'd need the MPU to place the kernel region towards the end of the allocated section, then move it further towards the front to achieve alignment, while not colliding with the app's memory, and then passing back the allocated _start_ address (which isn't even subject to that alignment) to the caller.

Apparently, even our current interface has sufficient complexity implement it incorrectly (#3544), and adding all of the above constraints adds so many more edge-cases.

I don't like either, but for now I'd be leaning towards allocating another word of memory, which can be used by grants at runtime anyways.

Reply:
Ok yes I agree. Given that we are allocating a larger region, it makes sense that the process standard impl just ensures it can both give the process enough accessible memory and make its grant bookkeeping aligned.